util/client/httpclient/response.go

```diff
@@ -0,0 +1,222 @@
+		restyResp = nil
+	}
+
+	if !restyResp.IsSuccess() && err == nil {
+		// Turn all non success responses to an error, like 301 Moved Permanently and 404 Not Found.
+		// Note: IsError != !IsSuccess.
+		err = ErrRequestFailed.New("%s %s (%s): Response status %d",
+			restyResp.Request.Method,
+			restyResp.Request.URL,
+			lResp.requestSnapshot.kindTag,
+			restyResp.StatusCode())
+	}
+
+	if err != nil {
+		// Turn response into nil when there is an error.
+		if restyResp != nil && restyResp.RawResponse != nil {
+			data, _ := ioutil.ReadAll(restyResp.RawResponse.Body)
+			_ = restyResp.RawResponse.Body.Close()
+			info.respStatus = restyResp.Status()
+			info.respBody = string(data)
+			restyResp = nil
+		}
+		info.Warn("Request failed", err)
+	}
+
+	if restyResp != nil && restyResp.RawResponse != nil {
+		lResp.executedResponseBody = restyResp.RawResponse.Body
+		lResp.executedResponseWithoutBody = restyResp.RawResponse
+		restyResp.RawResponse.Body = nil
+	} else {
+		lResp.executedResponseBody = nil
+		lResp.executedResponseWithoutBody = nil
+	}
+	lResp.executedError = err
+	lResp.executeInfo = info
+	lResp.isExecuted = true
+
+	// The request is executed, no need to schedule a check for the execution any more.
+	runtime.SetFinalizer(lResp, nil)
+}
+
+func (lResp *LazyResponse) close() {
+	_ = lResp.executedResponseBody.Close()
+}
+
+// Finish closes the response. Read is not possible any more.
+// The returned raw response does not have a body so that you don't need to close it manually.
+func (lResp *LazyResponse) Finish() (respNoBody *http.Response, err error) {
+	lResp.doExecutionOnce()
+	if lResp.executedError != nil {
+		return nil, lResp.executedError
+	}
+	respNoBody = lResp.executedResponseWithoutBody
+	lResp.close()
+	return
+}
+
+func (lResp *LazyResponse) PipeBody(w io.Writer) (written int64, respNoBody *http.Response, err error) {
+	lResp.doExecutionOnce()
+	if lResp.executedError != nil {
+		return 0, nil, lResp.executedError
+	}
+	respNoBody = lResp.executedResponseWithoutBody
+	written, err = io.Copy(w, lResp.executedResponseBody)
+	if err != nil {
+		respNoBody = nil
+		err = ErrRequestFailed.WrapWithNoMessage(err)
+		lResp.executeInfo.Warn("Request failed", err)
+	}
+	lResp.close()
+	return
+}
+
+func (lResp *LazyResponse) ReadBodyAsBytes() (bytes []byte, respNoBody *http.Response, err error) {
+	lResp.doExecutionOnce()
+	if lResp.executedError != nil {
+		return nil, nil, lResp.executedError
+	}
+	respNoBody = lResp.executedResponseWithoutBody
+	bytes, err = ioutil.ReadAll(lResp.executedResponseBody)
+	if err != nil {
+		bytes = nil
+		respNoBody = nil
+		err = ErrRequestFailed.WrapWithNoMessage(err)
+		lResp.executeInfo.Warn("Request failed", err)
+	}
+	lResp.close()
+	return
+}
+
+func (lResp *LazyResponse) ReadBodyAsString() (data string, respNoBody *http.Response, err error) {
+	bytes, resp, err := lResp.ReadBodyAsBytes()
+	if err != nil {
+		return "", nil, err
+	}
+	return strings.TrimSpace(string(bytes)), resp, nil
```

Comment:
Maybe it's too surprising to have `TrimSpace`, as a common http client.

Reply:
This behavior is taken from resty:

```go
// String method returns the body of the server response as String.
func (r *Response) String() string {
	if r.body == nil {
		return ""
	}
	return strings.TrimSpace(string(r.body))
}
```